x/gov/keeper/deposit.go

```diff
@@ -162,6 +163,73 @@ func (keeper Keeper) AddDeposit(ctx sdk.Context, proposalID uint64, depositorAdd
 	return activatedVotingPeriod, nil
 }
 
+// chargeDeposit will charge proposal cancellation fee (deposits * proposal_cancel_burn_rate)  and
+// send to a destAddress if defined or burn otherwise.
+// Remaining funds are send back to the depositor.
+func (keeper Keeper) chargeDeposit(ctx sdk.Context, proposalID uint64, destAddress, proposalCancelRate string, totalDeposits []sdk.Coin) error {
+	store := ctx.KVStore(keeper.storeKey)
+
+	rate := sdk.MustNewDecFromStr(proposalCancelRate)
+	var cancellationCharges sdk.Coins
+
+	if rate.IsPositive() {
+		for _, deposit := range totalDeposits {
+			burnAmount := sdk.NewCoin(
+				deposit.Denom,
+				sdk.NewDecFromInt(deposit.Amount).Mul(rate).RoundInt(),
+			)
+			cancellationCharges = cancellationCharges.Add(burnAmount)
+		}
+	}
+
+	// burn the cancellation fee or sent to cancellation changes to destination address
+	if !cancellationCharges.IsZero() {
+		// get the distribution module account address
+		distributionAddress := keeper.authKeeper.GetModuleAddress(disttypes.ModuleName)
+		if distributionAddress.String() == destAddress {
+			err := keeper.distrkeeper.FundCommunityPool(ctx, cancellationCharges, keeper.ModuleAccountAddress())
+			if err != nil {
+				return err
+			}
+		} else {
+			if len(destAddress) == 0 {
+				// burn the remaining deposits also
+				err := keeper.bankKeeper.BurnCoins(ctx, types.ModuleName, cancellationCharges)
+				if err != nil {
+					return err
+				}
+			} else {
+				destAccAddress := sdk.MustAccAddressFromBech32(destAddress)
+				err := keeper.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, destAccAddress, cancellationCharges)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	// send remaining (deposits - cancellationCharges) to depositors
+	for _, deposits := range keeper.GetDeposits(ctx, proposalID) {
+		depositerAddress := sdk.MustAccAddressFromBech32(deposits.Depositor)
+		var remainingAmount sdk.Coins
+		for _, deposit := range deposits.Amount {
+			remainAmount := sdk.NewCoin(
+				deposit.Denom,
+				sdk.NewDecFromInt(deposit.Amount).Sub(sdk.NewDec(1).Sub(rate)).RoundInt(),
```

Comment:
```suggestion
				sdk.NewDecFromInt(deposit.Amount).Mul(sdk.NewDec(1).Sub(rate)).RoundInt(),
```

Am I understanding this correctly?

Reply:
Sorry

Reply:
let's add a test for this case.

Reply:
Hmm, I added test cases here https://github.com/cosmos/cosmos-sdk/pull/13010/files#diff-72b620cb17246319713aafae5b0e10536e2ea6d5aefbe107ffc7982362b1a514R367-R374

Reply:
please add some unit tests as well?